### PR TITLE
Water pump 3 status

### PR DESCRIPTION
**Pull request by DamijanD**
Added: 
- Water pump 3 sensor (used with zone 2).
- Mixing valve step
- Mixing valve status

### CODE REVIEW

File: components/ecodan/status.h
Comment (gekkekoe) [inline components/ecodan/status.h:1]: you need to also add it to the query array  `statusCmdQueue`.

I don't have 2 zones so please test it carefully. I cannot test/verify it myself. Most ideal scenario is when another user also confirms it is working.

File: protocol.md
Comment (gekkekoe) [inline protocol.md:1]: Please try to keep the protocol in sync with @F1p repo. This doc change is probably not needed

**Review by gekkekoe — CHANGES_REQUESTED** [review]

File: protocol.md
Comment (F1p) [reply protocol.md:1]: > Please try to keep the protocol in sync with @F1p repo. This doc change is probably not needed

Agree, was going to keep pump generic because different configurations might use different pump outputs

File: components/ecodan/status.h
Comment (DamijanD) [reply components/ecodan/status.h:1]: Thx for promptly review. I would miss that query array 100% :-)

Yes I know that 2 zones are not so common. I'll do my best with testing.

File: protocol.md
Comment (DamijanD) [reply protocol.md:1]: Noted and reverted

File: components/ecodan/status.h
Comment (DamijanD) [reply components/ecodan/status.h:1]: @gekkekoe I guess I am stuck. So I see new sensor on the status page (IP of atom/esp) I also see it on the message pane (on the right) at the start (I guess binary sensors send message only at start and on change). 
The problem is that value is always 0 (OFF) - I waited more than 1 day just to be sure (pump gets stopped during DHW cycle) but there was no change... 
Just for test I changed the code in response.cpp to force ON and this showed me that this new case was executed. 
So question is if the  0x16 [2] really has the Out3 info as specified in the protocol?
Is there any possibility to get the dump of the whole message/package (without physical connection to the ESP)? If not as a workaround I could probably create new text_sensor and fill it with the res.debug_dump_packet()

Comment (F1p) [reply components/ecodan/status.h:1]: on my MQTT version of this project pump 3 is working for this location

image

Comment (DamijanD) [reply components/ecodan/status.h:1]: How is your HP actually set? Which FTC version (mine is 6)? Do you have 2 zones? And if not what is the function of that pump3?

Comment (F1p) [reply components/ecodan/status.h:1]: FTC5, pre-plumb cylinder - same as FTC6

Zone2 into OUT3
Zone1 into OUT13
OUT2 no connection (zone1)
DHW into CNP4/Pump 4
Then a primary circ pump too

File: components/ecodan/response.cpp
Comment (gekkekoe) [inline components/ecodan/response.cpp:192]: curious, why the || true ?

Comment (DamijanD) [reply components/ecodan/response.cpp]: That was just a test to see if that part of code was executed or not. (I mentioned it in https://github.com/gekkekoe/esphome-ecodan-hp/pull/73#discussion_r1852601160)

Comment (gekkekoe) [reply components/ecodan/response.cpp]: ah good, lets wait for 1 additional user report, and then I will merge it

Comment (DamijanD) [reply components/ecodan/response.cpp]: Not yet - now I am fetching both x15 & x16 responses and:
image
So it looks like that I get WaterPump3 status in byte 5 of x15 response - but I need to wait few hours to get to DHW cycle to confirm this.

Comment (gekkekoe) [reply components/ecodan/response.cpp]: you can increase debug level in yaml and read the logged msg :)

hmm byte 5 of 0x15. 0x16 is not reporting anything for you. @F1p 0x16 worked for you right?

Comment (F1p) [reply components/ecodan/response.cpp]: Procon confirms 0x15 bytes 4/5/6/7 matches Discrete Input (Digital Input):
image

I got strange results from holding registers though.

"Water Pump 1/2/3" may not link or match FTC board marking "OUT" 1/2/3 though...


I worked with a FTC6 user extensively on pumps and we looked at PCB outputs and linked them to bits.
My Zone2 pump and "status of heating" align with 0x16 byte 2 and the pump is connected to PCB marking "OUT3"
My DHW water pump is connected to CPN4 and aligns with 0x16 byte 1
image

Comment (F1p) [reply components/ecodan/response.cpp]: > Not yet - now I am fetching both x15 & x16 responses and: image So it looks like that I get WaterPump3 status in byte 5 of x15 response - but I need to wait few hours to get to DHW cycle to confirm this.

Can you trace the cable to see what "OUT" marking on the PCB is triggering for this byte please?

Comment (gekkekoe) [reply components/ecodan/response.cpp]: @DamijanD So per @F1p byte 5 is pump 3. Can you see where it's connected on your main board ? Is it connected to one of the OUT 1/2/3 sockets or something else ?

Comment (DamijanD) [reply components/ecodan/response.cpp]: I analysed the 0x15 response and found out this:
- Byte 5 - water pump 3 (out3) ON/OFF
- Byte 11 - Mixing valve  (out5) : 0:idle 1:closing 2:opening 3:close/stop
- Byte 10 - this has maybe something to do with mixing valve position (values 0, 8, 9, A) 

@F1p:
image
image

Comment (gekkekoe) [reply components/ecodan/response.cpp]: I asked @F1p to check it with the procon since I don't have a procon at my disposal. byte 5 is correct.
We can remove 0x16 from the query then, and update protocol.md.

If you also want the mixing valve status, that one can be added as well. Probably best to keep it numeric and translate it in the yaml.

btw, when you have some time, also have a look at: https://github.com/gekkekoe/esphome-ecodan-hp/discussions/70#discussioncomment-11349614

Comment (F1p) [reply components/ecodan/response.cpp]: Going to monitor my 0x15 byte 5 to compare with 0x16 byte 3


image

Comment (DamijanD) [reply components/ecodan/response.cpp]: Byte 3 is not static for me: (most of the time value is 1F)
RES{.Hdr{FC,62,02,7A,10}.Payload{15,01,1F,1F,01,01,00,01,00,00,0A,00,00,00,00,00}.Chk{B1}}
RES{.Hdr{FC,62,02,7A,10}.Payload{15,01,1F,20,01,01,00,01,00,00,09,00,00,00,00,00}.Chk{B1}}
RES{.Hdr{FC,62,02,7A,10}.Payload{15,01,1F,1E,00,00,01,00,01,01,00,03,00,00,00,00}.Chk{B9}}
RES{.Hdr{FC,62,02,7A,10}.Payload{15,01,1F,1D,00,00,01,00,00,00,00,00,00,00,00,00}.Chk{BF}}

@F1p Do you also have mixing valve for Zone2? I don't know how it can be set but in melcloud response there is `"HasSimplifiedZone2": false,` so maybe 0x15 is used for "proper" Zone2 and 0x16 for "simplified"  ?

Comment (F1p) [reply components/ecodan/response.cpp]: Never had MELCloud, but no mixing tank here so perhaps i do have simplified zone2..

File: components/ecodan/response.cpp
Comment (gekkekoe) [inline components/ecodan/response.cpp:193]: If we leave this as an internal variable, you can use something like https://github.com/gekkekoe/esphome-ecodan-hp/blob/5d0efaea72b55d9e758db12db55610d70f8ea90f/confs/base.yaml#L129 to map it internal value to a readable string

File: components/ecodan/sensor.py
Comment (gekkekoe) [inline components/ecodan/sensor.py:296]: see previous comment, this is probably not needed

File: components/ecodan/response.cpp
Comment (gekkekoe) [reply components/ecodan/response.cpp:193]: for 0x15, byte 11, other system are reporting 04 there.

Byte 11 - Mixing valve (out5) : 0:idle 1:closing 2:opening 3:close/stop
You did not see this value ?

File: components/ecodan/response.cpp
Comment (DamijanD) [reply components/ecodan/response.cpp]: Probably mixing tank is not what is important here but motorized mixing valve...

**Comment by gekkekoe** [comment]
It's looking good now. Remove the draft flag when you need me to do a final review.
Thanks for all the efforts so far.

File: components/ecodan/response.cpp
Comment (DamijanD) [reply components/ecodan/response.cpp]: I've created text sensor in confs/zone2.yaml (line 85) but for some reason it always returns "Idle" - currently I have both sensors and I see that value vas changed, but this text sensor was not updated...

Comment (DamijanD) [reply components/ecodan/response.cpp]: Hmm is it possible that "refresh rate" for this text sensor is too low:
image
The 3rd change was detected - but duration is for whole minute, and these corrections usually last for only about 10-15s. If this is the case then it would be best if we would send the text code directly from c++ part right?

Comment (gekkekoe) [reply components/ecodan/response.cpp]: Ah, so we only query 0x15 once every 21 seconds. So it would not matter if done via c++ directly. Since the 0x15 response is only retrieve once every 21 seconds. you can try to reduce the polling to 500ms (https://github.com/gekkekoe/esphome-ecodan-hp/blob/5d0efaea72b55d9e758db12db55610d70f8ea90f/components/ecodan/__init__.py#L27) but that would give you once every 10.5 seconds.

I do generally emulate what the procon and melcloud are doing, they also seem to query 1 topic per second, but we might need to tweak it

Comment (gekkekoe) [reply components/ecodan/response.cpp]: > I've created text sensor in confs/zone2.yaml (line 85) but for some reason it always returns "Idle" - currently I have both sensors and I see that value vas changed, but this text sensor was not updated...

you can try to publish the template sensor, it will trigger the rendering. you can do that when the value was changed

Comment (DamijanD) [reply components/ecodan/response.cpp]: That is not so important/critical property so I don't think that this justifies increased polling. Another example:
image

Comment (gekkekoe) [reply components/ecodan/response.cpp]: agreed, would like to keep it as close as the official adapters as possible

Comment (F1p) [reply components/ecodan/response.cpp]: > Hmm is it possible that "refresh rate" for this text sensor is too low: image The 3rd change was detected - but duration is for whole minute, and these corrections usually last for only about 10-15s. If this is the case then it would be best if we would send the text code directly from c++ part right?

I think this is similar to the defrost stages - i have run polling down to ~350ms stable per msg, but issues were reported before for some bytes slow

@DamijanD - Can you verify multi-zone working from 0x03 please:
image

Comment (DamijanD) [reply components/ecodan/response.cpp]: @F1p currently I get 1 which corresponds with real situation. I will keep an eye on it :-)

Comment (gekkekoe) [reply components/ecodan/response.cpp]: nice, but lets try not to add more scope to this PR :) Let's have this tested and merged.

Comment (F1p) [reply components/ecodan/response.cpp]: > Mixing valve

Byte 10 = Mixing Step in procon
image

Unable to locate Byte 11 currently

Comment (DamijanD) [reply components/ecodan/response.cpp]: > nice, but lets try not to add more scope to this PR :) Let's have this tested and merged.
Agree I cleaned up the code. (but I will make new PR with this change after this one is merged because it is something that is nice to have in multi zone setup).
I also removed the text part for mixing valve status as nothing I did was responsive enough (and we don't want to mapping to text status in response.cpp because of multi language).

Comment (gekkekoe) [reply components/ecodan/response.cpp]: new PR are always appreciated. And napping in c++ code would not make any difference. It's only refreshed from the heatpump once every 21s

**Pull request merged**
